File: iOS/DemoObjC/ScenarioListViewController.m
#import "ScenarioListViewController.h"
#import "ArticleViewController.h"
#import "DemoModels.h"
#import <MembranaAdWallSDK/MembranaAdWallSDK-Swift.h>

@interface ScenarioListViewController ()
@property (nonatomic, strong) MBRAdWall *adWall;
@property (nonatomic, strong) UIActivityIndicatorView *initializationIndicator;
@end

@implementation ScenarioListViewController

- (instancetype)initWithAdWall:(MBRAdWall *)adWall {
    self = [super initWithStyle:UITableViewStyleInsetGrouped];
    if (self) {
        _adWall = adWall;
        self.title = @"AdWall Scenarios";
    }
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    [self.tableView registerClass:UITableViewCell.class forCellReuseIdentifier:@"Article"];
    self.tableView.rowHeight = 118;
    self.navigationController.navigationBar.prefersLargeTitles = YES;
    self.initializationIndicator = [[UIActivityIndicatorView alloc] initWithActivityIndicatorStyle:UIActivityIndicatorViewStyleMedium];
    self.initializationIndicator.accessibilityLabel = @"Preparing ads";
    [self.initializationIndicator startAnimating];
    self.navigationItem.rightBarButtonItem = [[UIBarButtonItem alloc] initWithCustomView:self.initializationIndicator];
}

- (void)completeAdWallStartWithError:(NSError *)error {
    [self.initializationIndicator stopAnimating];
    self.navigationItem.rightBarButtonItem = nil;
    // Deferred one run loop turn: setting the prompt before the nav bar's own layout pass has
    // settled (e.g. right at launch) can briefly report a benign but noisy zero-width constraint
    // conflict on its internal prompt view.
    NSString *prompt = error ? @"Ads are currently unavailable" : nil;
    dispatch_async(dispatch_get_main_queue(), ^{
        self.navigationItem.prompt = prompt;
    });
}

+ (DemoArticleGateScope)scopeForRow:(NSInteger)row {
    switch (row) {
        case 0: return DemoArticleGateScopeWholePage;
        case 1: return DemoArticleGateScopePartialPage;
        default: return DemoArticleGateScopeBrandedAppearance;
    }
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return 3;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"Article" forIndexPath:indexPath];
    UIListContentConfiguration *content = [UIListContentConfiguration subtitleCellConfiguration];
    switch ([ScenarioListViewController scopeForRow:indexPath.row]) {
        case DemoArticleGateScopeWholePage:
            content.text = @"Whole page";
            content.secondaryText = @"Protect the complete article from the first frame";
            content.image = [UIImage systemImageNamed:@"rectangle.fill"];
            break;
        case DemoArticleGateScopePartialPage:
            content.text = @"Part of screen";
            content.secondaryText = @"Keep the introduction visible and gate the remaining article";
            content.image = [UIImage systemImageNamed:@"rectangle.split.2x1"];
            break;
        case DemoArticleGateScopeBrandedAppearance:
            content.text = @"Branded styling";
            content.secondaryText = @"Same whole-page coverage, restyled with publisher colors and fonts";
            content.image = [UIImage systemImageNamed:@"paintpalette"];
            break;
    }
    content.imageProperties.tintColor = UIColor.systemPinkColor;
    content.textProperties.numberOfLines = 2;
    content.secondaryTextProperties.numberOfLines = 2;
    cell.contentConfiguration = content;
    cell.accessoryType = UITableViewCellAccessoryDisclosureIndicator;
    cell.selectionStyle = UITableViewCellSelectionStyleDefault;
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    DemoArticleGateScope scope = [ScenarioListViewController scopeForRow:indexPath.row];
    UIViewController *controller = [[ArticleViewController alloc] initWithArticle:DemoContent.article
                                                                           adWall:self.adWall
                                                                        gateScope:scope];
    [self.navigationController pushViewController:controller animated:YES];
}

@end

File: iOS/DemoUIKit/ScenarioListViewController.swift
import MembranaAdWallSDK
import UIKit

final class ScenarioListViewController: UITableViewController {
    private enum Scenario: CaseIterable {
        case wholePage
        case partialPage
        case brandedAppearance

        var title: String {
            switch self {
            case .wholePage: return "Whole page"
            case .partialPage: return "Part of screen"
            case .brandedAppearance: return "Branded styling"
            }
        }

        var detail: String {
            switch self {
            case .wholePage: return "Protect the complete article from the first frame"
            case .partialPage: return "Keep the introduction visible and gate the remaining article"
            case .brandedAppearance: return "Same whole-page coverage, restyled with publisher colors and fonts"
            }
        }

        var symbol: String {
            switch self {
            case .wholePage: return "rectangle.fill"
            case .partialPage: return "rectangle.split.2x1"
            case .brandedAppearance: return "paintpalette"
            }
        }

        var gateScope: ArticleGateScope {
            switch self {
            case .wholePage: return .wholePage
            case .partialPage: return .partialPage
            case .brandedAppearance: return .brandedAppearance
            }
        }
    }

    private let adWall: MembranaAdWall
    private let initializationIndicator = UIActivityIndicatorView(style: .medium)

    init(adWall: MembranaAdWall) {
        self.adWall = adWall
        super.init(style: .insetGrouped)
        title = "AdWall Scenarios"
        navigationItem.largeTitleDisplayMode = .always
    }

    @available(*, unavailable)
    required init?(coder: NSCoder) { fatalError("init(coder:) has not been implemented") }

    override func viewDidLoad() {
        super.viewDidLoad()
        navigationController?.navigationBar.prefersLargeTitles = true
        tableView.register(UITableViewCell.self, forCellReuseIdentifier: "Scenario")
        tableView.rowHeight = 82
        tableView.accessibilityIdentifier = "scenario.list"
        initializationIndicator.accessibilityLabel = "Preparing test ads"
        initializationIndicator.startAnimating()
        navigationItem.rightBarButtonItem = UIBarButtonItem(customView: initializationIndicator)
    }

    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        Scenario.allCases.count
    }

    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let scenario = Scenario.allCases[indexPath.row]
        let cell = tableView.dequeueReusableCell(withIdentifier: "Scenario", for: indexPath)
        var content = cell.defaultContentConfiguration()
        content.text = scenario.title
        content.secondaryText = scenario.detail
        content.image = UIImage(systemName: scenario.symbol)
        content.imageProperties.tintColor = .systemPink
        cell.contentConfiguration = content
        cell.accessoryType = .disclosureIndicator
        cell.selectionStyle = .default
        return cell
    }

    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
        let scenario = Scenario.allCases[indexPath.row]
        let controller = ArticleViewController(
            article: DemoContent.article,
            adWall: adWall,
            gateScope: scenario.gateScope
        )
        navigationController?.pushViewController(controller, animated: true)
    }

    func completeAdWallStart(with error: Error?) {
        initializationIndicator.stopAnimating()
        navigationItem.rightBarButtonItem = nil
        demoSetPrompt(error == nil ? "Google test ads ready" : "Ads are currently unavailable")
    }
}
